net/portmapper/upnp.go

```diff
@@ -169,11 +192,63 @@ func getUPnPClient(ctx context.Context, gw netaddr.IP) (upnpClient, error) {
 		}
 	}()
 
-	select {
-	case client := <-clients:
-		return client, nil
-	case <-ctx.Done():
-		return nil, ctx.Err()
+	// In addition to pinging specific endpoints, if we are able to find any
+	discoClients := make(chan upnpClient, 3)
+	var total int32 = 6
+	var done int32
+	discoCtxt, discoCancel := context.WithTimeout(context.Background(), 5*time.Second)
+	go func() {
+		var err error
+		ip1Clients, _, err := internetgateway2.NewWANIPConnection1Clients(discoCtxt)
+		if err == nil && len(ip1Clients) > 0 {
+			discoClients <- ip1Clients[0]
+		}
+		if atomic.AddInt32(&done, 1) == total {
```

Comment:
can we use some different mechanism? this is a little magic and frail. why do we even need to close discoClients at all?